zeroex/orderwatch/order_watcher_test.go

```diff
@@ -0,0 +1,666 @@
+// +build !js
+
+package orderwatch
+
+import (
+	"context"
+	"flag"
+	"math/big"
+	"testing"
+	"time"
+
+	"github.com/0xProject/0x-mesh/constants"
+	"github.com/0xProject/0x-mesh/ethereum"
+	"github.com/0xProject/0x-mesh/ethereum/blockwatch"
+	"github.com/0xProject/0x-mesh/ethereum/dbstack"
+	"github.com/0xProject/0x-mesh/ethereum/wrappers"
+	"github.com/0xProject/0x-mesh/meshdb"
+	"github.com/0xProject/0x-mesh/scenario"
+	"github.com/0xProject/0x-mesh/zeroex"
+	"github.com/0xProject/0x-mesh/zeroex/ordervalidator"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/ethclient"
+	ethrpc "github.com/ethereum/go-ethereum/rpc"
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	ethereumRPCRequestTimeout   = 30 * time.Second
+	blockWatcherRetentionLimit  = 20
+	blockPollingInterval        = 1000 * time.Millisecond
+	ethereumRPCMaxContentLength = 524288
+)
+
+var makerAddress = constants.GanacheAccount1
+var takerAddress = constants.GanacheAccount2
+var eighteenDecimalsInBaseUnits = new(big.Int).Exp(big.NewInt(10), big.NewInt(18), nil)
+var wethAmount = new(big.Int).Mul(big.NewInt(50), eighteenDecimalsInBaseUnits)
+var zrxAmount = new(big.Int).Mul(big.NewInt(100), eighteenDecimalsInBaseUnits)
+var tokenID = big.NewInt(1)
+
+// Since these tests must be run sequentially, we don't want them to run as part of
+// the normal testing process. They will only be run if the "--serial" flag is used.
+var serialTestsEnabled bool
+
+func init() {
+	flag.BoolVar(&serialTestsEnabled, "serial", false, "enable serial tests")
+	flag.Parse()
+}
+
+var rpcClient *ethrpc.Client
+var ethClient *ethclient.Client
+var zrx *wrappers.ZRXToken
+var dummyERC721Token *wrappers.DummyERC721Token
+var exchange *wrappers.Exchange
+var weth *wrappers.WETH9
+
+func init() {
```

Comment:
Just a thought: I'm pretty comfortable with this `init` function so far because it purely does set up and should never fail under normal circumstances. However, we should keep an eye on it. We want to avoid testing conditions/assumptions/behaviors within `init` because the error messages can be harder to interpret. Anything that could reasonably fail should be included in a test function where we have access to `testing.T`.